File: utils/src/main/java/jetbrains/exodus/core/dataStructures/SoftObjectCacheBase.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.core.dataStructures;

import jetbrains.exodus.core.dataStructures.hash.HashUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.ref.SoftReference;

public abstract class SoftObjectCacheBase<K, V> extends ObjectCacheBase<K, V> {

    public static final int MIN_SIZE = 16;

    private final int chunkSize;
    private final SoftReference<ObjectCacheBase<K, V>>[] chunks;

    SoftObjectCacheBase(int cacheSize) {
        super(cacheSize);
        if (cacheSize < MIN_SIZE) {
            cacheSize = MIN_SIZE;
        }
        //noinspection unchecked
        chunks = new SoftReference[computeNumberOfChunks(cacheSize)];
        chunkSize = cacheSize / chunks.length;
        clear();
    }

    public void clear() {
        for (int i = 0; i < chunks.length; i++) {
            chunks[i] = null;
        }
    }

    @Override
    public final void lock() {
    }

    @Override
    public final void unlock() {
    }

    @Override
    public V tryKey(@NotNull final K key) {
        incAttempts();
        final ObjectCacheBase<K, V> chunk = getChunk(key, false);
        final V result = chunk == null ? null : chunk.tryKeyLocked(key);
        if (result != null) {
            incHits();
        }
        return result;
    }

    @Override
    public V getObject(@NotNull final K key) {
        final ObjectCacheBase<K, V> chunk = getChunk(key, false);
        if (chunk == null) {
            return null;
        }
        try (CriticalSection ignored = chunk.newCriticalSection()) {
            return chunk.getObject(key);
        }
    }

    @Override
    public V cacheObject(@NotNull final K key, @NotNull final V value) {
        final ObjectCacheBase<K, V> chunk = getChunk(key, true);
        assert chunk != null;
        try (CriticalSection ignored = chunk.newCriticalSection()) {
            return chunk.cacheObject(key, value);
        }
    }

    @Override
    public V remove(@NotNull final K key) {
        final ObjectCacheBase<K, V> chunk = getChunk(key, false);
        if (chunk == null) {
            return null;
        }
        try (CriticalSection ignored = chunk.newCriticalSection()) {
            return chunk.remove(key);
        }
    }

    @Override
    public int count() {
        throw new UnsupportedOperationException();
    }

    @Override
    public CriticalSection newCriticalSection() {
        return TRIVIAL_CRITICAL_SECTION;
    }

    static int computeNumberOfChunks(final int cacheSize) {
        int result = (int) Math.sqrt(cacheSize);
        while (result * result < cacheSize) {
            ++result;
        }
        return HashUtil.getCeilingPrime(result);
    }

    protected abstract ObjectCacheBase<K, V> newChunk(final int chunkSize);

    @Nullable
    private ObjectCacheBase<K, V> getChunk(@NotNull final K key, final boolean create) {
        final int hc = key.hashCode();
        final int chunkIndex = ((hc + (hc >> 31)) & 0x7fffffff) % chunks.length;
        final SoftReference<ObjectCacheBase<K, V>> ref = chunks[chunkIndex];
        ObjectCacheBase<K, V> result = ref == null ? null : ref.get();
        if (result == null && create) {
            result = newChunk(chunkSize);
            chunks[chunkIndex] = new SoftReference<>(result);
        }
        return result;
    }
}

File: utils/src/main/java/jetbrains/exodus/core/dataStructures/SoftObjectCacheBase.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.core.dataStructures

import jetbrains.exodus.core.dataStructures.hash.*
import java.lang.ref.SoftReference
import java.util.*

abstract class SoftObjectCacheBase<K, V> internal constructor(cacheSize: Int) : ObjectCacheBase<K, V>(cacheSize) {
    private val chunkSize: Int
    private val chunks: Array<SoftReference<ObjectCacheBase<K, V>?>>

    init {
        var cacheSize = cacheSize
        if (cacheSize < MIN_SIZE) {
            cacheSize = MIN_SIZE
        }
        chunks = arrayOfNulls<SoftReference<*>>(computeNumberOfChunks(cacheSize))
        chunkSize = cacheSize / chunks.size
        clear()
    }

    override fun clear() {
        Arrays.fill(chunks, null)
    }

    override fun lock() {}
    override fun unlock() {}
    override fun tryKey(key: K): V {
        incAttempts()
        val chunk = getChunk(key, false)
        val result = chunk?.tryKeyLocked(key)
        if (result != null) {
            incHits()
        }
        return result
    }

    override fun getObject(key: K): V {
        val chunk = getChunk(key, false) ?: return null
        chunk.newCriticalSection().use { ignored -> return chunk.getObject(key) }
    }

    override fun cacheObject(key: K, value: V): V {
        val chunk = getChunk(key, true)
        chunk!!.newCriticalSection().use { ignored -> return chunk.cacheObject(key, value) }
    }

    override fun remove(key: K): V {
        val chunk = getChunk(key, false) ?: return null
        chunk.newCriticalSection().use { ignored -> return chunk.remove(key) }
    }

    override fun count(): Int {
        throw UnsupportedOperationException()
    }

    override fun newCriticalSection(): CriticalSection? {
        return ObjectCacheBase.Companion.TRIVIAL_CRITICAL_SECTION
    }

    protected abstract fun newChunk(chunkSize: Int): ObjectCacheBase<K, V>
    private fun getChunk(key: K, create: Boolean): ObjectCacheBase<K, V>? {
        val hc = key.hashCode()
        val chunkIndex = (hc + (hc shr 31) and 0x7fffffff) % chunks.size
        val ref = chunks[chunkIndex]
        var result = if (ref == null) null else ref.get()
        if (result == null && create) {
            result = newChunk(chunkSize)
            chunks[chunkIndex] = SoftReference(result)
        }
        return result
    }

    companion object {
        const val MIN_SIZE = 16
        fun computeNumberOfChunks(cacheSize: Int): Int {
            var result = Math.sqrt(cacheSize.toDouble()).toInt()
            while (result * result < cacheSize) {
                ++result
            }
            return HashUtil.getCeilingPrime(result)
        }
    }
}
